h/search/index.py

```diff
@@ -66,6 +66,44 @@ def index(self, annotation_ids: list, windowsize: int = PG_WINDOW_SIZE):
                 errored.add(status["_id"])
         return errored
 
+    def delete(self, annotation_ids: list[str]) -> None:
+        """Delete `annotation_ids` from Elasticsearch."""
+        # Delete the given annotations from Elasticsearch by sending them to
+        # Elasticsearch's bulk API in chunks.
+        #
+        # Use streaming_bulk() rather than bulk() because of this comment in
+        # the Elasticsearch docs:
+        #
+        # > When errors are being collected original document data is included
+        # > in the error dictionary which can lead to an extra high memory
+        # > usage. If you need to process a lot of data and want to
+        # > ignore/collect errors please consider using the streaming_bulk()
+        # > helper which will just return the errors and not store them in
+        # > memory.
+        # >
+        # > https://elasticsearch-py.readthedocs.io/en/6.8.2/helpers.html#elasticsearch.helpers.bulk
+        results = es_helpers.streaming_bulk(
+            client=self.es_client.conn,
+            actions=[
+                {
+                    "_index": self._target_index,
+                    "_type": "annotation",
+                    "_id": annotation_id,
+                    "doc": {"deleted": True},
```

Comment:
This `{"deleted": True}` body that we use for deleted annotations in Elasticsearch is also [duplicated in `SearchIndexService.delete_annotation_by_id()`](https://github.com/hypothesis/h/blob/9686c96713812af3aead485129861072604756b0/h/services/search_index.py#L116). We might want to do something to remove that duplication.

Reply:
This `delete()` method has an odd behaviour: if you pass it an annotation ID that doesn't exist then it will add that annotation ID to the search index with the body `{deleted: true}`.

The [existing method for deleting annotations](https://github.com/hypothesis/h/blob/8c42be11a487a8fa6ce65637255404facd06c853/h/services/search_index.py#L101-L116) has the same issue.

This is because it's using `streaming_bulk()`'s `"_op_type": "index"` option (the default). You can pass `"op_type": "update"` instead but this will just add `deleted: true` into the existing document, it won't delete all the other fields. Or we could use `"_op_type": "delete"`, which is the right way to delete documents from Elasticsearch, but this actually deletes the documents rather than replacing them with `{deleted: true}` and I don't have time to address that issue with Hypothesis's behaviour right now.

The Elasticsearch docs for this are here: https://elasticsearch-py.readthedocs.io/en/6.8.2/helpers.html?highlight=streaming_bulk#bulk-helpers